src/internal/pfsdb/repos.go

```diff
@@ -0,0 +1,354 @@
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, ErrRepoNotFound{Project: repoProject, Name: repoName, Type: repoType}
+		}
+		return nil, errors.Wrap(err, "scanning repo row")
+	}
+	return row, nil
+}
+
+// todo(fahad): rewrite branch related code during the branches migration.
+// GetRepo retrieves an entry from the pfs.repos table by using the row id.
+func GetRepo(ctx context.Context, tx *pachsql.Tx, id pachsql.ID) (*pfs.RepoInfo, error) {
+	if id == 0 {
+		return nil, errors.New("invalid id: 0")
+	}
+	row := &repoRow{}
+	err := tx.QueryRowxContext(ctx, fmt.Sprintf("%s WHERE repo.id=$1 GROUP BY repo.id, project.name;", getRepoAndBranches), id).StructScan(row)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, ErrRepoNotFound{ID: id}
+		}
+		return nil, errors.Wrap(err, "scanning repo row")
+	}
+	return getRepoFromRepoRow(row)
+}
+
+// GetRepoByName retrieves an entry from the pfs.repos table by project, repo name, and type.
+func GetRepoByName(ctx context.Context, tx *pachsql.Tx, repoProject, repoName, repoType string) (*pfs.RepoInfo, error) {
+	row, err := getRepoRowByName(ctx, tx, repoProject, repoName, repoType)
+	if err != nil {
+		return nil, err
+	}
+	return getRepoFromRepoRow(row)
+}
+
+func getRepoFromRepoRow(row *repoRow) (*pfs.RepoInfo, error) {
+	proj := &pfs.Project{
+		Name: row.ProjectName,
+	}
+	branches, err := getBranchesFromRepoRow(row)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting branches from repo row")
+	}
+	repoInfo := &pfs.RepoInfo{
+		Repo: &pfs.Repo{
+			Name:    row.Name,
+			Type:    row.RepoType,
+			Project: proj,
+		},
+		Description: row.Description,
+		Branches:    branches,
+		Created:     timestamppb.New(row.CreatedAt),
+	}
+	return repoInfo, nil
+}
+
+func getBranchesFromRepoRow(row *repoRow) ([]*pfs.Branch, error) {
+	if row == nil {
+		return nil, errors.Wrap(fmt.Errorf("repo row is nil"), "")
+	}
+	branches := make([]*pfs.Branch, 0)
+	if row.Branches == noBranches {
+		return branches, nil
+	}
+	// after aggregation, braces, quotes, and leading hex prefixes need to be removed from the encoded branch strings.
+	for _, branchStr := range strings.Split(strings.Trim(row.Branches, "{}"), ",") {
+		branchHex := strings.Trim(strings.Trim(branchStr, "\""), "\\x")
+		decodedString, err := hex.DecodeString(branchHex)
+		if err != nil {
+			return nil, errors.Wrap(err, "branch not hex encoded")
+		}
+		branchInfo := &pfs.BranchInfo{}
+		if err := proto.Unmarshal(decodedString, branchInfo); err != nil {
+			return nil, errors.Wrap(err, "error unmarshalling")
+		}
+		branches = append(branches, branchInfo.Branch)
+	}
+	return branches, nil
+}
+
+// UpsertRepo updates all fields of an existing repo entry in the pfs.repos table by name. If 'upsert' is set to true, UpsertRepo()
+// will attempt to call CreateRepo() if the entry does not exist.
+func UpsertRepo(ctx context.Context, tx *pachsql.Tx, repo *pfs.RepoInfo) error {
```

Comment:
In Postgres, the canonical way of doing upserts is via the `ON CONFLICT` clause.

For example, lets say you already inserted `repo1` previously, you can upsert with the following:

```sql
INSERT INTO pfs.repos (name, type, project_id, description)
VALUES ('repo1', 'user', (SELECT id from core.projects where name='default'), 'repo1 description')
ON CONFLICT (name, type, project_id)
DO UPDATE SET description='updated description';
```

Note that we only update the `description` column because we already know for sure that the conflict is on `name, type, and project_id` column.

Reply:
Sounds good! I think the desired behavior we want is to allow the user to override everything except the ID if they are explicitly upserting. This would support future resource renames. What do you think?

Reply:
Your current upsert implementation doesn't support renaming a repo. nor should it. You are getting the repo ID by project, name, and type, and if there's an existing repo, then you update given that repo's ID. However, you already fixed the project, name, and type, so the only thing left to change is description. For renaming, your update functionality is all you need.

The correct semantics of upsert is to update an existing row if the value already exists, otherwise insert. To know if a row exists, we must identify it with a primary key or the unique id. We can't use id because we don't want to insert a custom id out of band, instead we want the db to manage the id to be internally consistent. Therefore, the only thing you can rely on is the primary key, which means the only column left that's available for updating is the description.